File: plugins/weather/weather.30m.dart
#!/usr/bin/env dart
/// Weather Plugin - Uses Crossbar web API
import 'dart:convert';
import 'dart:io';

String? crossbarWeb(String url) {
  try {
    final result = Process.runSync('crossbar', ['web', url]);
    if (result.exitCode == 0) {
      return (result.stdout as String).trim();
    }
  } catch (_) {}
  return null;
}

void main() async {
  final apiKey = Platform.environment['WEATHER_API_KEY'] ?? '';
  final city = Platform.environment['WEATHER_CITY'] ?? 'London';

  if (apiKey.isEmpty) {
    print('🌡️ No API Key');
    print('---');
    print('Set WEATHER_API_KEY');
    return;
  }

  final url = 'api.openweathermap.org/data/2.5/weather?q=$city&appid=$apiKey&units=metric';

  // Try Crossbar web first
  var response = crossbarWeb(url);

  // Fallback to HttpClient
  if (response == null || response.isEmpty) {
    try {
      final request = await HttpClient().getUrl(Uri.parse('https://$url'));
      final httpResponse = await request.close();
      response = await httpResponse.transform(utf8.decoder).join();
    } catch (_) {}
  }

  if (response != null && response.isNotEmpty) {
    try {
      final data = jsonDecode(response) as Map<String, dynamic>;
      final temp = data['main']?['temp'] ?? '--';
      final desc = (data['weather'] as List?)?.first?['description'] ?? '';
      
      print('🌡️ $temp°C');
      print('---');
      print('Location: $city');
      print('Temperature: $temp°C');
      print('Condition: $desc');
    } catch (_) {
      print('🌡️ Parse Error');
    }
  } else {
    print('🌡️ Error');
    print('---');
    print('Failed to fetch data');
  }
  
  print('---');
  print('Refresh | refresh=true');
}

File: plugins/weather/weather.30m.js
#!/usr/bin/env node
/**
 * Weather Plugin - Uses Crossbar web API
 */
const { execSync } = require('child_process');
const https = require('https');

function crossbarWeb(url) {
    try {
        return execSync(`crossbar web "${url}"`, { encoding: 'utf8', timeout: 10000 }).trim();
    } catch {
        return null;
    }
}

const API_KEY = process.env.WEATHER_API_KEY || '';
const CITY = process.env.WEATHER_CITY || 'London';

if (!API_KEY) {
    console.log('🌡️ No API Key');
    console.log('---');
    console.log('Set WEATHER_API_KEY');
    process.exit(0);
}

const url = `api.openweathermap.org/data/2.5/weather?q=${CITY}&appid=${API_KEY}&units=metric`;

// Try Crossbar web first
let response = crossbarWeb(url);

if (response) {
    try {
        const data = JSON.parse(response);
        const temp = data.main?.temp || '--';
        const desc = data.weather?.[0]?.description || '';
        
        console.log(`🌡️ ${temp}°C`);
        console.log('---');
        console.log(`Location: ${CITY}`);
        console.log(`Temperature: ${temp}°C`);
        console.log(`Condition: ${desc}`);
        console.log('---');
        console.log('Refresh | refresh=true');
    } catch {
        console.log('🌡️ Parse Error');
    }
} else {
    // Fallback to native https
    https.get(`https://${url}`, (res) => {
        let data = '';
        res.on('data', (chunk) => data += chunk);
        res.on('end', () => {
            try {
                const json = JSON.parse(data);
                const temp = json.main?.temp || '--';
                const desc = json.weather?.[0]?.description || '';
                
                console.log(`🌡️ ${temp}°C`);
                console.log('---');
                console.log(`Location: ${CITY}`);
                console.log(`Temperature: ${temp}°C`);
                console.log(`Condition: ${desc}`);
            } catch {
                console.log('🌡️ Parse Error');
            }
            console.log('---');
            console.log('Refresh | refresh=true');
        });
    }).on('error', () => {
        console.log('🌡️ Error');
        console.log('---');
        console.log('Failed to fetch data');
    });
}
